File: fe1-web/Model/Network/Method/Message/data/rollCall/closeRollCall.ts
import { Hash } from "Model/Objects/hash";
import { Timestamp } from "Model/Objects/timestamp";
import { ActionType, MessageData, ObjectType } from "../messageData";
import { PublicKey } from "Model/Objects/publicKey";

export class CloseRollCall implements MessageData {

    public readonly object: ObjectType = ObjectType.ROLL_CALL;
    public readonly action: ActionType = ActionType.CLOSE;

    public readonly id: Hash;
    public readonly start: Timestamp;
    public readonly end: Timestamp;
    public readonly attendees: PublicKey[];

    constructor(msg: Partial<CloseRollCall>) {
        Object.assign(this, msg);
    }

    public verify(): boolean {
        // to be implemented...
        return true;
    }
}

File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/rollCall/CloseRollCall.scala
package ch.epfl.pop.model.network.method.message.data.rollCall

import ch.epfl.pop.json.MessageDataProtocol._
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.{Hash, PublicKey, Timestamp}
import spray.json._

final case class CloseRollCall(
    update_id: Hash,
    closes: Hash,
    closed_at: Timestamp,
    attendees: List[PublicKey]
) extends MessageData {
  override val _object: ObjectType = ObjectType.roll_call
  override val action: ActionType = ActionType.close
}

object CloseRollCall extends Parsable {
  def apply(
      update_id: Hash,
      closes: Hash,
      closed_at: Timestamp,
      attendees: List[PublicKey]
  ): CloseRollCall = {
    new CloseRollCall(update_id, closes, closed_at, attendees)
  }

  override def buildFromJson(payload: String): CloseRollCall = payload.parseJson.asJsObject.convertTo[CloseRollCall]
}
